crates/fluvio/src/consumer/retry.rs

```diff
@@ -0,0 +1,279 @@
+                    Poll::Ready(None)
+                }
+                Poll::Pending => Poll::Pending,
+            }
+        } else {
+            Poll::Ready(None)
+        }
+    }
+}
+
+impl ConsumerStream for ConsumerWithRetry {
+    fn offset_commit(&mut self) -> std::result::Result<(), ErrorCode> {
+        if let Some(ref mut stream) = self.current_stream {
+            stream.offset_commit()
+        } else {
+            warn!("Can't commit, no current stream");
+            Ok(())
+        }
+    }
+
+    fn offset_flush(
+        &mut self,
+    ) -> futures_util::future::BoxFuture<'_, std::result::Result<(), ErrorCode>> {
+        if let Some(ref mut stream) = self.current_stream {
+            stream.offset_flush()
+        } else {
+            warn!("Can't flush, no current stream");
+            Box::pin(async { Ok(()) })
+        }
+    }
+}
+
+impl ConsumerWithRetry {
+    /// Creates a new `ConsumerWithRetry` instance.
+    pub async fn new(fluvio_config: FluvioConfig, config: ConsumerConfigExt) -> Result<Self> {
+        let fluvio = Fluvio::connect_with_config(&fluvio_config).await?;
+        let stream = fluvio.consumer_with_config_inner(config.clone()).await?;
+        let boxed_stream: BoxConsumerStream = Box::pin(stream);
+
+        Ok(Self {
+            inner: ConsumerWithRetryInner {
+                fluvio_client: Arc::new(fluvio),
+                next_offset_to_read: None,
+                consumer_config: config,
+            },
+            current_stream: Some(boxed_stream),
+            pending: None,
+        })
+    }
+
+    /// Consumes records from the stream with retry logic.
+    ///
+    /// This function continuously attempts to retrieve a record from the current stream.
+    /// On failure (or end-of-stream), it waits using an exponential backoff strategy
+    /// and then attempts to reconnect.
+    async fn consumer_with_retry(
+        inner: ConsumerWithRetryInner,
+        fluvio_client: Arc<Fluvio>,
+        mut current_stream: BoxConsumerStream,
+    ) -> Option<Result<(ConsumerRecord, BoxConsumerStream, Option<i64>), ErrorCode>> {
+        let mut attempts: u32 = 0;
+        let mut backoff = create_backoff();
+
+        loop {
+            // Try to retrieve the next record.
+            if let Some(record_result) = current_stream.as_mut().next().await {
+                match record_result {
+                    Ok(record) => {
+                        let new_offset = Some(record.offset + 1);
+                        if attempts > 0 {
+                            debug!(
+                                target: SPAN_RETRY,
+                                "Record produced successfully after reconnect"
+                            );
+                        }
+                        return Some(Ok((record, current_stream, new_offset)));
+                    }
+                    Err(e) => {
+                        warn!(target: SPAN_RETRY, "Error consuming record: {}", e);
```

Comment:
return error?

Reply:
we can't return error here, only if retry mode is Disabled. We need to keep trying on next lines of code.